File: syntavra_runtime/broker_live_snapshot_contract.py
from __future__ import annotations

import hashlib
import os
import sqlite3
import stat
import time
from pathlib import Path
from typing import Any, Final

from .broker_snapshot_contract import (
    BROKER_SCHEMA_VERSION,
    CONTRACT_VERSION,
    DATABASE_NAME,
    TABLE_SPECS,
    BrokerSnapshotError,
    _broker_schema_version,
    _canonical_json_bytes,
    _database_path,
    _foreign_keys,
    _indexes,
    _project_root,
    _schema_objects,
    _table_columns,
    _table_rows,
)

SNAPSHOT_SCHEMA_VERSION: Final = 1
SNAPSHOT_ID: Final = "syntavra-broker-live-snapshot-v1"
MAXIMUM_DATABASE_BYTES: Final = 64 * 1024 * 1024
MAXIMUM_DURATION_SECONDS: Final = 5.0
PAGES_PER_STEP: Final = 64
RETRY_SLEEP_SECONDS: Final = 0.010
_SIDECAR_SUFFIXES: Final = ("-journal", "-shm", "-wal")


class BrokerLiveSnapshotError(BrokerSnapshotError):
    pass


def _error(code: str, cause: BaseException | None = None) -> BrokerLiveSnapshotError:
    error = BrokerLiveSnapshotError(code)
    if cause is not None:
        error.__cause__ = cause
    return error


def _sidecar_path(database: Path, suffix: str) -> Path:
    return Path(f"{database}{suffix}")


def _regular_file_metadata(path: Path, code_prefix: str) -> os.stat_result | None:
    try:
        metadata = path.lstat()
    except FileNotFoundError:
        return None
    except OSError as exc:
        raise _error(f"{code_prefix}_METADATA_FAILED", exc)
    if stat.S_ISLNK(metadata.st_mode):
        raise _error(f"{code_prefix}_SYMLINK")
    if not stat.S_ISREG(metadata.st_mode):
        raise _error(f"{code_prefix}_NOT_FILE")
    return metadata


def _sidecar_state(database: Path) -> dict[str, bool]:
    state = {
        "journal": _regular_file_metadata(
            _sidecar_path(database, "-journal"),
            "BROKER_LIVE_SIDECAR",
        )
        is not None,
        "shm": _regular_file_metadata(
            _sidecar_path(database, "-shm"),
            "BROKER_LIVE_SIDECAR",
        )
        is not None,
        "wal": _regular_file_metadata(
            _sidecar_path(database, "-wal"),
            "BROKER_LIVE_SIDECAR",
        )
        is not None,
    }
    if state["journal"]:
        raise _error("BROKER_LIVE_ROLLBACK_JOURNAL_PRESENT")
    if state["wal"] != state["shm"]:
        raise _error("BROKER_LIVE_WAL_SHM_PAIR_INVALID")
    return state


def _observed_identity(database: Path, sidecars: dict[str, bool]) -> tuple[tuple[str, int, int], ...]:
    paths = [("database", database)]
    for name in ("shm", "wal"):
        if sidecars[name]:
            paths.append((name, _sidecar_path(database, f"-{name}")))
    output: list[tuple[str, int, int]] = []
    for name, path in paths:
        metadata = _regular_file_metadata(path, "BROKER_LIVE_SOURCE")
        if metadata is None:
            raise _error("BROKER_LIVE_SOURCE_DISAPPEARED")
        output.append((name, int(metadata.st_size), int(metadata.st_mtime_ns)))
    return tuple(output)


def _open_source(database: Path) -> sqlite3.Connection:
    uri = f"{database.as_uri()}?mode=ro"
    connection: sqlite3.Connection | None = None
    try:
        connection = sqlite3.connect(
            uri,
            uri=True,
            timeout=0.0,
            isolation_level=None,
        )
        connection.row_factory = sqlite3.Row
        connection.execute("PRAGMA query_only=ON")
        connection.execute("PRAGMA trusted_schema=OFF")
        query_only = connection.execute("PRAGMA query_only").fetchone()
    except sqlite3.Error as exc:
        if connection is not None:
            connection.close()
        raise _error("BROKER_LIVE_SOURCE_OPEN_FAILED", exc)
    if query_only is None or int(query_only[0]) != 1:
        connection.close()
        raise _error("BROKER_LIVE_QUERY_ONLY_FAILED")
    return connection


def _positive_pragma(connection: sqlite3.Connection, name: str) -> int:
    try:
        row = connection.execute(f"PRAGMA {name}").fetchone()
    except sqlite3.Error as exc:
        raise _error("BROKER_LIVE_SOURCE_METADATA_FAILED", exc)
    if row is None:
        raise _error("BROKER_LIVE_SOURCE_METADATA_FAILED")
    try:
        value = int(row[0])
    except (TypeError, ValueError) as exc:
        raise _error("BROKER_LIVE_SOURCE_METADATA_FAILED", exc)
    if value <= 0:
        raise _error("BROKER_LIVE_SOURCE_METADATA_FAILED")
    return value


def _journal_mode(connection: sqlite3.Connection) -> str:
    try:
        row = connection.execute("PRAGMA journal_mode").fetchone()
    except sqlite3.Error as exc:
        raise _error("BROKER_LIVE_SOURCE_METADATA_FAILED", exc)
    if row is None or not isinstance(row[0], str):
        raise _error("BROKER_LIVE_SOURCE_METADATA_FAILED")
    value = row[0].casefold()
    if value not in {"delete", "memory", "off", "persist", "truncate", "wal"}:
        raise _error("BROKER_LIVE_JOURNAL_MODE_UNSUPPORTED")
    return value


def _validate_size(page_size: int, page_count: int) -> int:
    if page_size <= 0 or page_size > 65536 or page_size & (page_size - 1):
        raise _error("BROKER_LIVE_PAGE_SIZE_INVALID")
    if page_count <= 0:
        raise _error("BROKER_LIVE_PAGE_COUNT_INVALID")
    logical_bytes = page_size * page_count
    if logical_bytes > MAXIMUM_DATABASE_BYTES:
        raise _error("BROKER_LIVE_DATABASE_TOO_LARGE")
    return logical_bytes


def _logical_snapshot(
    connection: sqlite3.Connection,
    expected_project_id: str,
) -> tuple[int, dict[str, list[dict[str, Any]]]]:
    connection.row_factory = sqlite3.Row
    try:
        connection.execute("PRAGMA query_only=ON")
        connection.execute("PRAGMA trusted_schema=OFF")
    except sqlite3.Error as exc:
        raise _error("BROKER_LIVE_DESTINATION_QUERY_ONLY_FAILED", exc)
    _schema_objects(connection)
    for table, spec in TABLE_SPECS.items():
        _table_columns(connection, table, spec[0])
    _indexes(connection)
    _foreign_keys(connection)
    schema_version = _broker_schema_version(connection)
    tables = {
        table: _table_rows(connection, table, spec, expected_project_id)
        for table, spec in TABLE_SPECS.items()
    }
    return schema_version, tables


def snapshot_live_broker_database(
    project_root: str | os.PathLike[str],
    database_path: str | os.PathLike[str],
    *,
    expected_project_id: str,
) -> dict[str, Any]:
    root, actual_project_id = _project_root(project_root, expected_project_id)
    database, relative_path = _database_path(root, database_path)
    sidecars_before = _sidecar_state(database)
    identity_before = _observed_identity(database, sidecars_before)

    source = _open_source(database)
    destination = sqlite3.connect(":memory:", isolation_level=None)
    started = time.monotonic()
    deadline = started + MAXIMUM_DURATION_SECONDS
    progress_state = {"steps": 0, "remaining": 0, "total": 0}
    try:
        journal_mode = _journal_mode(source)
        page_size = _positive_pragma(source, "page_size")
        initial_page_count = _positive_pragma(source, "page_count")
        initial_logical_bytes = _validate_size(page_size, initial_page_count)

        def progress(status: int, remaining: int, total: int) -> None:
            del status
            progress_state["steps"] += 1
            progress_state["remaining"] = int(remaining)
            progress_state["total"] = int(total)
            if time.monotonic() > deadline:
                raise _error("BROKER_LIVE_BACKUP_TIMEOUT")
            if total <= 0 or remaining < 0 or remaining > total:
                raise _error("BROKER_LIVE_BACKUP_PROGRESS_INVALID")
            _validate_size(page_size, int(total))

        try:
            source.backup(
                destination,
                pages=PAGES_PER_STEP,
                progress=progress,
                name="main",
                sleep=RETRY_SLEEP_SECONDS,
            )
        except BrokerLiveSnapshotError:
            raise
        except sqlite3.Error as exc:
            raise _error("BROKER_LIVE_BACKUP_FAILED", exc)
        if time.monotonic() > deadline:
            raise _error("BROKER_LIVE_BACKUP_TIMEOUT")
        if progress_state["steps"] <= 0 or progress_state["remaining"] != 0:
            raise _error("BROKER_LIVE_BACKUP_INCOMPLETE")

        final_page_count = _positive_pragma(destination, "page_count")
        final_logical_bytes = _validate_size(page_size, final_page_count)
        broker_schema_version, tables = _logical_snapshot(
            destination,
            expected_project_id,
        )
    finally:
        destination.close()
        source.close()

    sidecars_after = _sidecar_state(database)
    identity_after = _observed_identity(database, sidecars_after)
    source_changed = identity_after != identity_before or sidecars_after != sidecars_before

    row_counts = {table: len(rows) for table, rows in tables.items()}
    payload: dict[str, Any] = {
        "ok": True,
        "schema_version": SNAPSHOT_SCHEMA_VERSION,
        "contract_version": CONTRACT_VERSION,
        "snapshot_id": SNAPSHOT_ID,
        "broker_schema_version": broker_schema_version,
        "project_id": actual_project_id,
        "project_binding": {
            "expected": expected_project_id,
            "actual": actual_project_id,
            "matched": True,
        },
        "database": {
            "relative_path": relative_path,
            "source_open_mode": "read-only-live",
            "source_query_only": True,
            "source_journal_mode": journal_mode,
            "wal_present": sidecars_before["wal"],
            "shm_present": sidecars_before["shm"],
            "rollback_journal_present": False,
            "source_changed_during_backup": source_changed,
            "backup_destination": "memory",
        },
        "backup": {
            "api": "sqlite-online-backup",
            "pages_per_step": PAGES_PER_STEP,
            "maximum_database_bytes": MAXIMUM_DATABASE_BYTES,
            "maximum_duration_milliseconds": int(MAXIMUM_DURATION_SECONDS * 1000),
            "retry_sleep_milliseconds": int(RETRY_SLEEP_SECONDS * 1000),
            "page_size": page_size,
            "initial_page_count": initial_page_count,
            "initial_logical_bytes": initial_logical_bytes,
            "final_page_count": final_page_count,
            "final_logical_bytes": final_logical_bytes,
            "steps": progress_state["steps"],
            "complete": True,
        },
        "tables": tables,
        "row_counts": row_counts,
        "mutation": {
            "source_connection_writes": False,
            "checkpoint": False,
            "vacuum": False,
            "migration": False,
            "destination_files": False,
            "persistent_project_state": False,
        },
        "claim": "RUST_BROKER_SQLITE_LIVE_BACKUP_PARITY_PROVEN_R10_FIXTURES",
    }
    payload["snapshot_hash"] = hashlib.sha256(_canonical_json_bytes(payload)).hexdigest()
    return payload


__all__ = [
    "BrokerLiveSnapshotError",
    "MAXIMUM_DATABASE_BYTES",
    "MAXIMUM_DURATION_SECONDS",
    "PAGES_PER_STEP",
    "RETRY_SLEEP_SECONDS",
    "SNAPSHOT_ID",
    "snapshot_live_broker_database",
]

File: crates/syntavra-cli/src/broker_live_snapshot_contract.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::thread;
use std::time::{Duration, Instant, SystemTime};

use rusqlite::backup::{Backup, StepResult};
use rusqlite::{Connection, OpenFlags};
use serde_json::{Map, Number, Value};
use syntavra_core::sha256_hex;

use crate::broker_snapshot_contract::{
    broker_schema_version, canonical_project_root, contract as logical_contract, contract_array,
    contract_string, contract_u64, foreign_keys, indexes, percent_encode_path,
    relative_database_path, schema_objects, table_columns, table_rows,
};

const CONTRACT_JSON: &str = include_str!("../../../contracts/state/broker-live-snapshot-v1.json");
const DATABASE_NAME: &str = "broker.sqlite3";
const MAXIMUM_DATABASE_BYTES: u64 = 64 * 1024 * 1024;
const MAXIMUM_DURATION_MILLISECONDS: u64 = 5_000;
const PAGES_PER_STEP: i32 = 64;
const RETRY_SLEEP_MILLISECONDS: u64 = 10;

#[derive(Debug, Clone, PartialEq, Eq)]
struct ObservedIdentity {
    files: BTreeMap<String, (u64, Option<SystemTime>)>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct SidecarState {
    journal: bool,
    shm: bool,
    wal: bool,
}

fn error(code: &str) -> String {
    code.to_owned()
}

fn contract() -> Result<Value, String> {
    serde_json::from_str(CONTRACT_JSON).map_err(|_| error("BROKER_LIVE_CONTRACT_INVALID"))
}

fn regular_file(path: &Path, prefix: &str) -> Result<Option<fs::Metadata>, String> {
    match fs::symlink_metadata(path) {
        Ok(metadata) => {
            if metadata.file_type().is_symlink() {
                return Err(error(&format!("{prefix}_SYMLINK")));
            }
            if !metadata.is_file() {
                return Err(error(&format!("{prefix}_NOT_FILE")));
            }
            Ok(Some(metadata))
        }
        Err(source) if source.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(_) => Err(error(&format!("{prefix}_METADATA_FAILED"))),
    }
}

fn sidecar_path(database: &Path, suffix: &str) -> Result<PathBuf, String> {
    let value = database
        .to_str()
        .ok_or_else(|| error("BROKER_DATABASE_PATH_UTF8_INVALID"))?;
    Ok(PathBuf::from(format!("{value}{suffix}")))
}

fn sidecar_state(database: &Path) -> Result<SidecarState, String> {
    let state = SidecarState {
        journal: regular_file(
            &sidecar_path(database, "-journal")?,
            "BROKER_LIVE_SIDECAR",
        )?
        .is_some(),
        shm: regular_file(&sidecar_path(database, "-shm")?, "BROKER_LIVE_SIDECAR")?
            .is_some(),
        wal: regular_file(&sidecar_path(database, "-wal")?, "BROKER_LIVE_SIDECAR")?
            .is_some(),
    };
    if state.journal {
        return Err(error("BROKER_LIVE_ROLLBACK_JOURNAL_PRESENT"));
    }
    if state.wal != state.shm {
        return Err(error("BROKER_LIVE_WAL_SHM_PAIR_INVALID"));
    }
    Ok(state)
}

fn observed_identity(database: &Path, sidecars: SidecarState) -> Result<ObservedIdentity, String> {
    let mut paths = vec![("database", database.to_path_buf())];
    if sidecars.shm {
        paths.push(("shm", sidecar_path(database, "-shm")?));
    }
    if sidecars.wal {
        paths.push(("wal", sidecar_path(database, "-wal")?));
    }
    let mut files = BTreeMap::new();
    for (name, path) in paths {
        let metadata = regular_file(&path, "BROKER_LIVE_SOURCE")?
            .ok_or_else(|| error("BROKER_LIVE_SOURCE_DISAPPEARED"))?;
        files.insert(name.to_owned(), (metadata.len(), metadata.modified().ok()));
    }
    Ok(ObservedIdentity { files })
}

fn open_source(database: &Path) -> Result<Connection, String> {
    let uri = format!("file:{}?mode=ro", percent_encode_path(database)?);
    let flags = OpenFlags::SQLITE_OPEN_READ_ONLY
        | OpenFlags::SQLITE_OPEN_URI
        | OpenFlags::SQLITE_OPEN_NO_MUTEX;
    let connection = Connection::open_with_flags(uri, flags)
        .map_err(|_| error("BROKER_LIVE_SOURCE_OPEN_FAILED"))?;
    connection
        .busy_timeout(Duration::ZERO)
        .map_err(|_| error("BROKER_LIVE_SOURCE_OPEN_FAILED"))?;
    connection
        .execute_batch("PRAGMA query_only=ON; PRAGMA trusted_schema=OFF;")
        .map_err(|_| error("BROKER_LIVE_QUERY_ONLY_FAILED"))?;
    let query_only: i64 = connection
        .query_row("PRAGMA query_only", [], |row| row.get(0))
        .map_err(|_| error("BROKER_LIVE_QUERY_ONLY_FAILED"))?;
    if query_only != 1 {
        return Err(error("BROKER_LIVE_QUERY_ONLY_FAILED"));
    }
    Ok(connection)
}

fn positive_pragma(connection: &Connection, name: &str) -> Result<u64, String> {
    let sql = format!("PRAGMA {name}");
    let value: i64 = connection
        .query_row(&sql, [], |row| row.get(0))
        .map_err(|_| error("BROKER_LIVE_SOURCE_METADATA_FAILED"))?;
    u64::try_from(value)
        .ok()
        .filter(|value| *value > 0)
        .ok_or_else(|| error("BROKER_LIVE_SOURCE_METADATA_FAILED"))
}

fn journal_mode(connection: &Connection) -> Result<String, String> {
    let value: String = connection
        .query_row("PRAGMA journal_mode", [], |row| row.get(0))
        .map_err(|_| error("BROKER_LIVE_SOURCE_METADATA_FAILED"))?;
    let normalized = value.to_ascii_lowercase();
    if !matches!(
        normalized.as_str(),
        "delete" | "memory" | "off" | "persist" | "truncate" | "wal"
    ) {
        return Err(error("BROKER_LIVE_JOURNAL_MODE_UNSUPPORTED"));
    }
    Ok(normalized)
}

fn validate_size(page_size: u64, page_count: u64) -> Result<u64, String> {
    if page_size == 0 || page_size > 65_536 || !page_size.is_power_of_two() {
        return Err(error("BROKER_LIVE_PAGE_SIZE_INVALID"));
    }
    if page_count == 0 {
        return Err(error("BROKER_LIVE_PAGE_COUNT_INVALID"));
    }
    let logical_bytes = page_size
        .checked_mul(page_count)
        .ok_or_else(|| error("BROKER_LIVE_DATABASE_TOO_LARGE"))?;
    if logical_bytes > MAXIMUM_DATABASE_BYTES {
        return Err(error("BROKER_LIVE_DATABASE_TOO_LARGE"));
    }
    Ok(logical_bytes)
}

fn logical_snapshot(
    connection: &Connection,
    expected_project_id: &str,
) -> Result<(u64, Map<String, Value>, Map<String, Value>), String> {
    connection
        .execute_batch("PRAGMA query_only=ON; PRAGMA trusted_schema=OFF;")
        .map_err(|_| error("BROKER_LIVE_DESTINATION_QUERY_ONLY_FAILED"))?;
    let logical = logical_contract()?;
    schema_objects(connection, &logical)?;
    for table in contract_array(&logical, "tables")? {
        table_columns(connection, table)?;
    }
    indexes(connection, &logical)?;
    foreign_keys(connection, &logical)?;
    let schema_version = broker_schema_version(connection, &logical)?;
    let mut tables = Map::new();
    let mut row_counts = Map::new();
    for table in contract_array(&logical, "tables")? {
        let name = contract_string(table, "name")?;
        let rows = table_rows(connection, table, expected_project_id)?;
        row_counts.insert(name.to_owned(), Value::Number(Number::from(rows.len())));
        tables.insert(name.to_owned(), Value::Array(rows));
    }
    Ok((schema_version, tables, row_counts))
}

fn insert_project_binding(
    payload: &mut Map<String, Value>,
    expected_project_id: &str,
    actual_project_id: &str,
) {
    let mut binding = Map::new();
    binding.insert(
        "actual".to_owned(),
        Value::String(actual_project_id.to_owned()),
    );
    binding.insert(
        "expected".to_owned(),
        Value::String(expected_project_id.to_owned()),
    );
    binding.insert("matched".to_owned(), Value::Bool(true));
    payload.insert("project_binding".to_owned(), Value::Object(binding));
}

/// Produces a canonical logical snapshot from a bounded in-memory online backup.
///
/// # Errors
///
/// Returns a stable `BROKER_*` error code when source binding, sidecar policy,
/// backup bounds, SQLite online backup, or R9 logical validation fails.
#[allow(clippy::too_many_lines)]
pub fn snapshot_live_broker_database_json(
    project_root: &str,
    database_path: &str,
    expected_project_id: &str,
) -> Result<String, String> {
    let live = contract()?;
    if contract_string(&live, "database_name")? != DATABASE_NAME {
        return Err(error("BROKER_LIVE_CONTRACT_INVALID"));
    }
    if contract_u64(&live, "broker_schema_version")? != 2 {
        return Err(error("BROKER_LIVE_CONTRACT_INVALID"));
    }

    let (root, actual_project_id) = canonical_project_root(project_root, expected_project_id)?;
    let (database, relative_path) = relative_database_path(&root, database_path)?;
    let sidecars_before = sidecar_state(&database)?;
    let identity_before = observed_identity(&database, sidecars_before)?;
    let source = open_source(&database)?;
    let journal = journal_mode(&source)?;
    let page_size = positive_pragma(&source, "page_size")?;
    let initial_page_count = positive_pragma(&source, "page_count")?;
    let initial_logical_bytes = validate_size(page_size, initial_page_count)?;

    let mut destination = Connection::open_in_memory()
        .map_err(|_| error("BROKER_LIVE_DESTINATION_OPEN_FAILED"))?;
    let started = Instant::now();
    let deadline = Duration::from_millis(MAXIMUM_DURATION_MILLISECONDS);
    let backup = Backup::new(&source, &mut destination)
        .map_err(|_| error("BROKER_LIVE_BACKUP_FAILED"))?;
    let mut steps = 0_u64;
    loop {
        if started.elapsed() > deadline {
            return Err(error("BROKER_LIVE_BACKUP_TIMEOUT"));
        }
        steps = steps
            .checked_add(1)
            .ok_or_else(|| error("BROKER_LIVE_BACKUP_PROGRESS_INVALID"))?;
        let result = backup
            .step(PAGES_PER_STEP)
            .map_err(|_| error("BROKER_LIVE_BACKUP_FAILED"))?;
        let progress = backup.progress();
        if progress.pagecount <= 0
            || progress.remaining < 0
            || progress.remaining > progress.pagecount
        {
            return Err(error("BROKER_LIVE_BACKUP_PROGRESS_INVALID"));
        }
        validate_size(
            page_size,
            u64::try_from(progress.pagecount)
                .map_err(|_| error("BROKER_LIVE_BACKUP_PROGRESS_INVALID"))?,
        )?;
        match result {
            StepResult::Done => break,
            StepResult::More => {}
            StepResult::Busy | StepResult::Locked => {
                thread::sleep(Duration::from_millis(RETRY_SLEEP_MILLISECONDS));
            }
            _ => return Err(error("BROKER_LIVE_BACKUP_FAILED")),
        }
    }
    let progress = backup.progress();
    if progress.remaining != 0 {
        return Err(error("BROKER_LIVE_BACKUP_INCOMPLETE"));
    }
    drop(backup);

    if started.elapsed() > deadline {
        return Err(error("BROKER_LIVE_BACKUP_TIMEOUT"));
    }
    let final_page_count = positive_pragma(&destination, "page_count")?;
    let final_logical_bytes = validate_size(page_size, final_page_count)?;
    let (schema_version, tables, row_counts) =
        logical_snapshot(&destination, expected_project_id)?;
    drop(destination);
    drop(source);

    let sidecars_after = sidecar_state(&database)?;
    let identity_after = observed_identity(&database, sidecars_after)?;
    let source_changed = sidecars_after != sidecars_before || identity_after != identity_before;

    let mut database_value = Map::new();
    database_value.insert(
        "backup_destination".to_owned(),
        Value::String("memory".to_owned()),
    );
    database_value.insert(
        "relative_path".to_owned(),
        Value::String(relative_path),
    );
    database_value.insert(
        "rollback_journal_present".to_owned(),
        Value::Bool(false),
    );
    database_value.insert("shm_present".to_owned(), Value::Bool(sidecars_before.shm));
    database_value.insert(
        "source_changed_during_backup".to_owned(),
        Value::Bool(source_changed),
    );
    database_value.insert(
        "source_journal_mode".to_owned(),
        Value::String(journal),
    );
    database_value.insert(
        "source_open_mode".to_owned(),
        Value::String("read-only-live".to_owned()),
    );
    database_value.insert("source_query_only".to_owned(), Value::Bool(true));
    database_value.insert("wal_present".to_owned(), Value::Bool(sidecars_before.wal));

    let mut backup_value = Map::new();
    backup_value.insert(
        "api".to_owned(),
        Value::String("sqlite-online-backup".to_owned()),
    );
    backup_value.insert("complete".to_owned(), Value::Bool(true));
    backup_value.insert(
        "final_logical_bytes".to_owned(),
        Value::Number(Number::from(final_logical_bytes)),
    );
    backup_value.insert(
        "final_page_count".to_owned(),
        Value::Number(Number::from(final_page_count)),
    );
    backup_value.insert(
        "initial_logical_bytes".to_owned(),
        Value::Number(Number::from(initial_logical_bytes)),
    );
    backup_value.insert(
        "initial_page_count".to_owned(),
        Value::Number(Number::from(initial_page_count)),
    );
    backup_value.insert(
        "maximum_database_bytes".to_owned(),
        Value::Number(Number::from(MAXIMUM_DATABASE_BYTES)),
    );
    backup_value.insert(
        "maximum_duration_milliseconds".to_owned(),
        Value::Number(Number::from(MAXIMUM_DURATION_MILLISECONDS)),
    );
    backup_value.insert(
        "page_size".to_owned(),
        Value::Number(Number::from(page_size)),
    );
    backup_value.insert(
        "pages_per_step".to_owned(),
        Value::Number(Number::from(PAGES_PER_STEP)),
    );
    backup_value.insert(
        "retry_sleep_milliseconds".to_owned(),
        Value::Number(Number::from(RETRY_SLEEP_MILLISECONDS)),
    );
    backup_value.insert("steps".to_owned(), Value::Number(Number::from(steps)));

    let mut mutation = Map::new();
    mutation.insert("checkpoint".to_owned(), Value::Bool(false));
    mutation.insert("destination_files".to_owned(), Value::Bool(false));
    mutation.insert("migration".to_owned(), Value::Bool(false));
    mutation.insert("persistent_project_state".to_owned(), Value::Bool(false));
    mutation.insert("source_connection_writes".to_owned(), Value::Bool(false));
    mutation.insert("vacuum".to_owned(), Value::Bool(false));

    let mut payload = Map::new();
    payload.insert("backup".to_owned(), Value::Object(backup_value));
    payload.insert(
        "broker_schema_version".to_owned(),
        Value::Number(Number::from(schema_version)),
    );
    payload.insert(
        "claim".to_owned(),
        Value::String("RUST_BROKER_SQLITE_LIVE_BACKUP_PARITY_PROVEN_R10_FIXTURES".to_owned()),
    );
    payload.insert(
        "contract_version".to_owned(),
        Value::Number(Number::from(contract_u64(&live, "contract_version")?)),
    );
    payload.insert("database".to_owned(), Value::Object(database_value));
    payload.insert("mutation".to_owned(), Value::Object(mutation));
    payload.insert("ok".to_owned(), Value::Bool(true));
    insert_project_binding(&mut payload, expected_project_id, &actual_project_id);
    payload.insert("project_id".to_owned(), Value::String(actual_project_id));
    payload.insert("row_counts".to_owned(), Value::Object(row_counts));
    payload.insert(
        "schema_version".to_owned(),
        Value::Number(Number::from(contract_u64(&live, "schema_version")?)),
    );
    payload.insert(
        "snapshot_id".to_owned(),
        Value::String(contract_string(&live, "snapshot_id")?.to_owned()),
    );
    payload.insert("tables".to_owned(), Value::Object(tables));

    let encoded = serde_json::to_vec(&Value::Object(payload.clone()))
        .map_err(|_| error("BROKER_ROW_TYPE_INVALID"))?;
    payload.insert(
        "snapshot_hash".to_owned(),
        Value::String(sha256_hex(&encoded)),
    );
    serde_json::to_string(&Value::Object(payload)).map_err(|_| error("BROKER_ROW_TYPE_INVALID"))
}

#[cfg(test)]
mod tests {
    use super::{contract, snapshot_live_broker_database_json};

    #[test]
    fn embeds_the_r10_contract() {
        let value = contract().expect("contract");
        assert_eq!(value["snapshot_id"], "syntavra-broker-live-snapshot-v1");
        assert_eq!(value["backup_policy"]["pages_per_step"], 64);
    }

    #[test]
    fn rejects_invalid_expected_project_id_before_open() {
        assert_eq!(
            snapshot_live_broker_database_json(".", "broker.sqlite3", "invalid"),
            Err("BROKER_EXPECTED_PROJECT_INVALID".to_owned())
        );
    }
}
